### ISSUE TITLE
Elements objects cannot be serialised to JSON with standard JSON libraries

### ISSUE THREAD

**Issue by radumg** [hypar-io/Elements#130]
Went down a rabbit hole when one of our services was returning inexplicably truncated JSON responses, below is the developer's journey of debugging this, please bear with me @ikeough  🤓 

### In Microservices land 
We have our own object data model for a `Building` and an endpoint in a service (ASP .NET Core 2.2) that generates a new one.

Using the default serialiser settings that come with ASP .NET Core,  was finding that JSON responses are truncated producing invalid responses.

StackOverflow suggested using these options on the serialiser
```options.SerializerSettings.ReferenceLoopHandling = Newtonsoft.Json.ReferenceLoopHandling.Ignore```, but this makes the service go into a loop and the response is a nice infinite JSON. At this point I thought our own objects were at fault, no exceptions are thrown to hint that it goes into a self-referencing loop when serialising.

### In libraries
Moved to testing this in a library context. Added `JsonIgnore` attributes for computed properties that referenced another, that wasn't it.

Finally, when serialising using `Newtonsoft`, we get this error :
```
Self referencing loop detected for property 'Face' with type 'Elements.Geometry.Solids.Face'. Path 'Floors[0].FloorZone.geometry[0].Faces.0.Outer'.
```
Getting somewhere 😥 

Our `Building` model has an `List<Elements.Floor>` property, which in turn references a `Mass` which in turn has `Solids`, which has `Face`s that have `Loop`s that reference their parent faces.

![image](https://user-images.githubusercontent.com/11439624/56049843-37637380-5d42-11e9-965c-4cd3bf49a377.png)

Reading through the source code of `Hypar.Elements`, we find out why :
https://github.com/hypar-io/elements/blob/fa9353cb1e183595a222436f01ce24a53b3493bb/src/Elements/Serialization/JSON/SolidConverter.cs#L15-L19

### Custom converters
Trying to add the Hypar `SolidConverter` to our serialiser settings is a no-go either, since its constructor wants a weird dictionary of `<long, Material>`  - I have no way of producing that (unless a very expensive crawl of entire object tree keeping track of materials counts ).

In my opinion we should not have to use custom converters for these base types. 
It's unclear to me why references to parent objects are stored as `public properties` on `Loop` (or any other type) instead of these being private or behind `GetParent()` methods.

The `SolidConverter` itself seems to ignore these properties anyway when deserialising a model, which means they are not essential and can be reconstructed/computed.

I might be missing an obvious use case - just can't think of one.

### Proposal
Eliminate self-referencing properties in `Elements` so that the library can be used with standard serialisation libraries. 

The logic of all the custom converters will need to be replicated in languages other than C# when dealing with JSON files coming from `Hypar.Elements` when you don't use Hypar's own C# serialisation library.

Any thoughts ?

**Comment by radumg** [comment]
update : passing in an empty dictionary to the converter seems to work
```cs
var converter = new SolidConverter(new Dictionary<long, Elements.Material>())
```
however, looking at the number of custom converters available, it's probably only a matter of time before another type blows up due to my hacks ? 😁

**Comment by ikeough** [comment]
I feel your pain @radumg. It's been my experience however that JSON serialization is very rarely free. We use custom converters for a couple of purposes. The most important one is for referencing by id instead of serializing a reference into an object. For example, if a beam has the same profile as all the other beams, I don't want to serialize that profile every time. I want to create an object in the JSON called `Profiles` and serialize each Profile only once, then reference it by id. As you might imagine, this makes the JSON substantially smaller. Imagine if every Element had all of its `Material`, `ElementType`, and `Profile` properties serialized to each object. Not only would the JSON be massive, but when you deserialized your model you'd construct duplicates of all that stuff because the serializer doesn't know you're talking about the same thing.

The second reason you use a custom converter is because JSON deserialization doesn't work for inherited types without some type information. If a `Beam` inherits from `StructuralFraming`, which inherits from `Element`, the deserializer doesn't know how to try to deserialize to a `Beam`. We write a `type` field that contains the type information so we can instruct the serializer about the target type. Th way we do this is, by the way, only slightly different from how Json.net does this natively, except they write the type information into a field called `$type`. I just don't want `$type` in my JSON so I do this with a custom converter (`ElementConverter`). 

The third reason for converters, and one that I hadn't anticipated but am happy to have, is to allow deserialization of third-party types that derive from built in types. This was raised by @wynged.

Now, the specific case around geometry serialization is an interesting one. The solution you propose to remove the `Face` property on `Loop` is a possibility. Have you tried that though? My guess is that the problem is more pervasive than that because, as I say in that comment, BREPs are self-referencing structures at a number of levels to allow for navigation of the data structure to/from any entity. If we enact the "replace with a method" solution, we'll have to do this everywhere that this holds, which is possible, albeit a little bit unintuitive. A user of the library wants to do `someHalfEdge.Loop.Face` not `someHalfEdge.GetLoop().GetFace()`. Also, the`Face` property on `Loop` has a setter because it's set after the fact. So that setting either has to happen in something like a converter, or you have to set via the constructor in which case the value of `Face` has to be available to the serializer when it tries to deserialize the `Loop`.

**Comment by ikeough** [comment]
Also, the material dictionary hack will most likely become moot in the future because I'm going to take the `Material` property off of `Solid`. It turns out that the concept of a material in our domain is more nuanced than just a color and a glossiness. If you look at `IfcMaterial` as an example, you get into issues about specifying the layers and directionality of materials. This kind of stuff needs to happen at the `Element` level.

**Comment by ikeough** [comment]
Finally, I hear you about what this means for deserialization in other languages, and it's the reason why polyglot serialization frameworks, like Protocol Buffers, severely limit the kinds of nested data structures and inherited stuff that you can encode. Because decoding then requires custom decoding logic which is different in every language.

**Comment by ikeough** [comment]
One strategy for limiting the number of converters would be to change fields which reference things like `Profile`, `ElementType`, and `Material` to `long`, and just store ids. This would get rid of all the `<someType>ToId` converters. But then an element needs to have a reference to the `Model` which contains the collections of those things in order to access them. And I would prefer we not do that.

As a small improvement, `ElementTypeToIdConverter`, `ProfileToIdConverter` and `MaterialToIdConverter` can be consolidated into one generic class. They're all doing effectively the same thing, converting a reference to an `IIdentifiable` and back.

**Comment by radumg** [comment]
> I feel your pain @radumg
Thanks for the detailed response @ikeough , really appreciate it! Hope the details were useful as context and that it didn't read as complaining too much, not my intention 😇 

> In the mean time, the dictionary that you need is simply the `Model.Materials` dictionary.
Cool, will give this a go.

Agree with your 3 points actually, which kinda makes me contradict myself a bit 😁 :

> The most important one is for referencing by id instead of serializing a reference into an object. 
Makes sense to normalize in the context of a `Model`, we had more loosely-connected elements so didn't identify this as an issue at first.

> The second reason is because JSON deserialization doesn't work for inherited types without some type information. I just don't want `$type` in my JSON so I do this with a custom converter (`ElementConverter`).
Agree, we've been avoiding this too, especially since we're using a NoSQL document db.

> The third reason for converters, and one that I hadn't anticipated but am happy to have, is to allow deserialization of third-party types that derive from built in types. This was raised by @wynged.

We're inheriting from `Element` so this is also much appreciated.

**Comment by radumg** [comment]
> Now, the specific case around geometry serialization is an interesting one. The solution you propose to remove the `Face` property on `Loop` is a possibility. Have you tried that though? 

Not yet, but i'll give it a go. 

> we'll have to do this everywhere that this holds, which is possible, albeit a little bit unintuitive. A user of the library wants to do `someHalfEdge.Loop.Face` not `someHalfEdge.GetLoop().GetFace()`

I've stayed as far away from BREPs as I could so far, so can't speak to what's expected in this regard. No simple or singular answer to this by any means, but as a dev coming at this fresh though, this was more un-intuitive to me, always resulting in circular references (it's fun to be in a loop of `Peek definition in VS`). 

In the example above, that `someHalfEdge` wouldn't be pulled from thin air I assume, but its reference obtained from its parent somehow because data isn't normalised in a model here. So journey in code actually is
Solid > Faces[] > Face > Loop (outer) > Edges (List<HalfEdge>) > **HalfEdge** > Parent (back to Loop) > Parent (back to Face)

From **HalfEdge**, I'd find it reasonable to call `GetParent()` to go back up. Could even have nicely declared capabilities like `IHasParent<T>` & `IHasChildren<T>` which could help programmatically going up/down a dependency tree. All this feels more like a graph.

https://github.com/hypar-io/elements/blob/fa9353cb1e183595a222436f01ce24a53b3493bb/src/Elements/Geometry/Solids/Face.cs#L37
This feels like an unexpected side-effect, making it `outer.SetParent(this)` or `outer.SetFaceRef(this)` would feel more like recording relationships than modifying the related element. 

Guess the bits I'm struggling to grasp are (which may be to limited experience with this)
- use-case of why you'd need parent `Face` from a `HalfEdge` 
- why you wouldn't already have a reference to parent Face of Loop considering the above > flow 

> Also, the`Face` property on `Loop` has a setter because it's set after the fact. 
> or you have to set via the constructor in which case the value of `Face` has to be available to the serializer when it tries to deserialize the `Loop`.

`HalfEdge` already has internal constructors to do some of this, if parent refs were parameters on public constructors for all these classes (like DI), it would be more consistent: https://github.com/hypar-io/elements/blob/fa9353cb1e183595a222436f01ce24a53b3493bb/src/Elements/Geometry/Solids/HalfEdge.cs#L36

### plaster the issue idea
Just thinking out loud, but what if at least known circular references in classes were marked with`[JsonIgnore]` ? Assuming no impact on import/export, could solve the serialisation pain in the short-term.

### whacky idea
Most libraries I've seen leave it to you as a dev or a DI container to track these dependencies, so perhaps `Elements` can simplify its own API by not providing these refs, or only doing it when in the context of a `Model` ? (not even sure this can be achieved without this wider container, be it DI or a `Model` or other mechanisms.

**Comment by radumg** [comment]
> But then an element needs to have a reference to the `Model` which contains the collections of those things in order to access them

I agree requiring a `Model` to work with the `Element` classes isn't nice - especially since we're not using a `Model` at the minute. 

What if it was an optional param/method on constructors and these reference methods like `GetFace()` would only ever work if the context `Model` is set ?

**Comment by ikeough** [comment]
@radumg In our last discussion you mentioned challenges around serialization and CosmosDB. Can you elaborate? I’d like to know the bounds of the issue before I propose a change in how we serialize models.

**Comment by radumg** [comment]
I’m away this week but @alvpivkmans would be best places to provide some more context since he fixed the bug on our side. 

Alvaro, perhaps we could copy-paste some of the error messages/screenshots from our internal tracker to help illustrate the issues encountered  ?

**Comment by radumg** [comment]
Hi Ian,

Here's some details from one of our internal issues illustrating a serialisation issue, when using Azure Cosmos DB's (with Mongo API).

We create a `Profile` from front-end Vue app (to represent a site permiter)
![image](https://user-images.githubusercontent.com/11439624/60809655-149afc80-a183-11e9-9bac-b796c218e3ca.png)

The back-end API receives this successfully and deserialises it
![image](https://user-images.githubusercontent.com/11439624/60809746-4d3ad600-a183-11e9-9be3-7ec10743a3bc.png)

When the object lands in Cosmos DB however, the profile is `blank`/`null`
![image](https://user-images.githubusercontent.com/11439624/60809800-73607600-a183-11e9-9dc2-37cdef0e8577.png)

To get around this, we've had to replace `Profile` type used for the perimeter with our own `Boundary` class, which wraps around a Hypar `Polygon` and recreates it at deserialisation time from the list of `Vertices`.

```csharp
public class Boundary
    {
        private Polygon polygon { get; set; }
        public Vector3[] Vertices
        {
            get => polygon.Vertices;
            set => polygon = new Polygon(value);
        }

        public double Area => polygon.Area;
        public double Perimeter => polygon.Length();
        public Boundary(Vector3[] vertices)
        {
            this.Vertices = vertices;
        }
    }
```

**Comment by radumg** [comment]
Another example (bigger issue this time), with Hypar `Floors` refusing to serialise :

Steps :
- We have created JsonConverters that handle how hypar elements serilalise
- CosmosDB's MongoDB API doesn't care about those and simply serialises using a default (unknown) configuration, meaning that once we store them, all the geometrical data for floors is gone

![image](https://user-images.githubusercontent.com/11439624/60810423-01892c00-a185-11e9-8216-f09e3918e4be.png)

**Comment by radumg** [comment]
Hopefully the above 2 help, let us know if you have more questions or anything else we can supply 😄

**Comment by ikeough** [comment]
@radumg It's been a long time since I've reviewed this. In the interim we changed elements to use a different serializer to support our work on user-defined types for Hypar. I'm not sure if you'd had a chance to see what we've done or use >0.4.0. Have you had a chance to review what we've done with JSON schemas and are you using a newer version of Elements? If so, is this still an issue?

**Comment by ikeough** [comment]
Elements no longer serializes Solid geometry as referenced in the original issue. Closing this for now, but feel free to reopen if this problem persists.
